components/fxa-client/src/auth.rs

```diff
@@ -168,6 +157,28 @@ pub struct AuthorizationInfo {
     pub active: bool,
 }
 
+/// OAuth handler.  These are defined in the foreign code
+pub trait OAuthHandler: Send + Sync {
+    /// Perform an OAuth flow at a URL
+    ///
+    /// When the resulting OAuth flow redirects back to the configured `redirect_uri`,
+    /// the query parameters should be extracting from the URL and returned.
+    ///
+    /// Warning: the `FirefoxAccount` instance will be in the `Authorizing` state while this
+    /// method is running.  Consumers must make sure the method eventually returns or the
+    /// `FirefoxAccount` instance will be stuck.  Return `FxaError::Cancelled` for abandoned OAuth
+    /// sessions.
+    fn perform_flow(&self, url: String) -> CallbackResult<OAuthResult>;
+}
+
+// Result of an Oauth flow
+//
+// Normally, the field values are extracted from the URL query parameters when the browser reaches
+// the redirect_uri.
+pub struct OAuthResult {
+    pub code: String,
```

Comment:
I didn't realize this before, but this change means the consumer only needs to pass us the `code` query param, not the `state` anymore.  That makes sense though, since once we have a single method that handles the oauth process, we no longer need to round-trip data through the FxA server to get the oauth state setup at the begining.

Reply:
The `state` is used to verify that the authorization code returned to us by the server is the one associated with the same request we sent (see https://datatracker.ietf.org/doc/html/rfc6749#section-10.12)

We should still get the `state` back from the consumer and match it against the one we generated. If they don't match we should error out before sending the authorization request to the user

Reply:
Is this an issue for us though?  AIUI, the attack is against a stateless webserver where you redirect a user to the oauth complete URL, but with your code rather than theirs.  This doesn't seem possible because we control the entire OAuth flow inside our method.  When `OauthHandler::perform_oauth()` returns, I think we can be trust the code.  Also, if an attacker was somehow redirecting things inside the flow, they could probably see the state.

I'm not really sure about this though, so I'm very happy to put the state back in with a note that explains it's just in case.

Reply:
I definitely don't have the right knowledge in security or the OAuth 2.0 RFC itself so take everything I say with a grain of salt 😅 

I think you may be right about our case within the browsers, where the `code` and `state` are passed through a web channel configured **only** for the firefox.com origin. So the risk comes in:
- if the content server is compromised but there are worse things that will happen in that case (attackers gets user's credentials etc).
- Somehow the consuming apps try to complete the auth process by getting a `code` and `state` from a different source.. this implies a very bad bug in Firefox

So for Firefox it'll be precautionary and just in case (but I still think it's good to have regardless)

For hypothetical non-Firefox consumers, the issue becomes more apparent, we can imagine VPN or relay starting an oauth flow, where the OAuth handler blocks a thread, and resumes the thread with a `code` and `state` once a web redirect is done. If a third-party web application redirects to VPN or relay and passes their own `code` then it might not be possible for the relying party to confirm that it was indeed the oauth code it requested

Regardless, I think we should error on the side of caution here (I'm also not sure if the FxA server enforces a `state` query parameter or treats it as optional so we might be forced to keep it even if we didn't want to)

Reply:
This kinda ties into my comment above - if it's possible for multiple oauth requests to be happening concurrently, you can imagine a confused client causing something like this

Reply:
I'm still not sure this is needed, even in that VPN situation.  ISTM that if some attacker got in the middle of the flow, then they would be able to see the `state` and send it back.  The param is there to protect against a complete oauth endpoint from receiving the result of an an oauth flow initiated somewhere else.

But I'm not a security expert, so let's put it back.  I do think we can kill the `flow_store` field though.

Reply:
Just for completeness: My concern is that the delivery mechanism some other consumers might use to deliver the `code` and `state` may not have the same properties Firefox does. I could imagine a world where VPN uses [web redirect to redirect to its native app](https://www.oauth.com/oauth2-servers/redirect-uris/redirect-uris-native-apps/), and a third-party can take advantage of that to deliver its own oauth code.

I.e the concern is less that an attacker gets in the middle of the flow, it's more that an attacker inserts itself to the end of the flow and the relying party resolves the callback with that OAuth code.

But of course, I'm not a security expert either so +1 to putting it back if only for the reason of keeping the status quo here

Reply:
I updated the commit to still pass the `state` query, but removes the `flow_store` field.  I think think this should be very safe and avoids the issues of a confused client (if they mix up the requests, then they'll get 2 errors).